### PR TITLE
Create login page.

### PR DESCRIPTION
**Pull request by alex-benton**
Allows a user to log in and log out of the app.

They can't do anything else yet.

### CODE REVIEW

File: src/client/js/app.jsx [hunk 1]
```diff
@@ -1,12 +1,27 @@
 // @flow
-import React from 'react';
+import * as React from 'react';
```
Comment (alisazhou) [inline src/client/js/app.jsx:2]: is this for flow?

File: src/client/js/state/auth.js [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+ * - storing and retrieving the token we get
+ * - clearing auth data when a logout request is received
+ */
+import gql from 'graphql-tag';
+import { Mutation } from 'react-apollo';
+import * as React from 'react';
+import { Fragment } from 'react';
+import { Redirect } from 'react-router-dom';
+
+const jwtDecode = require('jwt-decode');
```
Comment (alisazhou) [inline src/client/js/state/auth.js:14]: can this be an es6 import?

File: src/client/js/state/auth.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+     *
+     * Update the token in localstorage, too.
+     *
+     * @param token the token to set
+     */
+    set token(token: ?string) {
+        this._token = token;
+        if (token) {
+            localStorage.setItem('token', token);
+            let decoded = jwtDecode(token);
```
Comment (alisazhou) [inline src/client/js/state/auth.js:43]: why `let` and not `const` here?

File: src/client/js/state/auth.js [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    render() {
+        let { children, ...rest } = this.props;
+        return !this.state.isAuthed ? (
+            <Redirect
+                to={{
+                    pathname: '/',
+                }}
+            />
+        ) : (
+            children(this.logout, rest)
```
Comment (alisazhou) [inline src/client/js/state/auth.js:167]: whoa til

File: src/client/js/components/Common/Header.jsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+const Header = props => <StateContext.Consumer>{state => <HeaderComponent {...props} />}</StateContext.Consumer>;
+
+/**
+ * An extremely-WIP header component.
+ */
+const HeaderComponent = props => (
+    <Fragment>
+        <Container>
+            <Logo color={'white'} size={45} />
+            <Logout>{logout => <button onClick={logout}>Log out</button>}</Logout>
```
Comment (alisazhou) [inline src/client/js/components/Common/Header.jsx:22]: the `rest` from [this line](https://github.com/frogrock-net/housekeeper/pull/10/files#diff-6b7e44e22608b239fa4c3db372e5cc8aR167) is ignored here right? and the args here are positional?
so if i need to i could do `{(logout, rest) => ...}`?

File: src/client/js/app.jsx [hunk 6]
```diff
@@ -1,12 +1,27 @@
...
 /**
  * The highest-level component for the housekeeper application.
  *
  * @constructor
  */
-const App = () => <Home />;
+const App = () => (
+    <StateContext.Provider value={state}>
+        <ApolloProvider client={state.client}>
+            <Route path={'/dashboard'} render={props => <RequireAuth {...props}>{props => <Dashboard {...props} />}</RequireAuth>} />
```
Comment (alisazhou) [inline src/client/js/app.jsx:21]: where is `RequireAuth` receiving `props.referrer`?

File: src/client/js/components/Common/Form/Form.jsx [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+     * Construct a form and initialize the state.
+     */
+    constructor(props: Props) {
+        super(props);
+
+        if (props.defaultValues) {
+            this.state = props.defaultValues;
+        } else {
+            this.state = {};
+        }
```
Comment (alisazhou) [inline src/client/js/components/Common/Form/Form.jsx:42]: `this.state = props.defaultValues || {}`?

File: src/client/js/components/Common/Form/Form.jsx [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+     * @param value the updated value
+     */
+    onUpdate = (field: string, value: any) => {
+        this.setState({ [field]: value });
+    };
+
+    /**
+     * Function that's executed when this form is submitted.
+     *
+     * NYI: form validation will occur here.
```
Comment (alisazhou) [inline src/client/js/components/Common/Form/Form.jsx:60]: New York Islanders?

File: src/client/js/components/Common/Form/Form.jsx [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    /**
+     * Render the Form component.
+     */
+    render() {
+        return (
+            <FormContainer className={this.props.className}>
+                <form onSubmit={this.onSubmit}>
+                    {React.Children.map(this.props.children, child => {
+                        return React.cloneElement(child, {
+                            onUpdate: this.onUpdate.bind(this),
```
Comment (alisazhou) [inline src/client/js/components/Common/Form/Form.jsx:78]: more til.

also, `this` here is `Form` right? isn't `onUpdate` already bound to `this`?

File: src/client/js/components/Common/Form/FormInput.jsx [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+    };
+
+    /**
+     * Update this component's state when it's focused.
+     *
+     * We track whether the input is focused in order to change the icon color along with the form field color.
+     *
+     * @param e the event. unused.
+     */
+    onFocus = (e: SyntheticInputEvent<HTMLInputElement>) => {
```
Comment (alisazhou) [inline src/client/js/components/Common/Form/FormInput.jsx:72]: does flow complain if you just do `onFocus = () => ...`?

File: src/client/js/state/house.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+type Props = {
+    children: CreateHouseRender,
+};
+
+/**
+ * Create house mutator component.
+ *
+ * @param children a render function
+ * @returns {*}
+ * @constructor
```
Comment (alisazhou) [inline src/client/js/state/house.js:66]: what does `@constructor` in the docstring do?

File: src/server/js/server.js [hunk 12]
```diff
@@ -84,6 +84,10 @@ app.listen(PORT, () => {
 
 let mongoose = require('mongoose');
 
+mongoose.Types.ObjectId.prototype.valueOf = function() {
+    return this.toString();
+};
```
Comment (alisazhou) [inline src/server/js/server.js:89]: this would be called like `room.id.valueOf()` eg?

**Review by alisazhou — COMMENTED** [review]
a bunch of questions

File: src/client/js/app.jsx (hunk 6 above)
Comment (alex-benton) [reply src/client/js/app.jsx:21]: it isn't yet

File: src/client/js/components/Common/Form/Form.jsx (hunk 7 above)
Comment (alex-benton) [reply src/client/js/components/Common/Form/Form.jsx:42]: ok

File: src/client/js/app.jsx (hunk 1 above)
Comment (alex-benton) [reply src/client/js/app.jsx:2]: yup

File: src/client/js/components/Common/Form/Form.jsx (hunk 9 above)
Comment (alex-benton) [reply src/client/js/components/Common/Form/Form.jsx:78]: yup

**Review by alisazhou — APPROVED** [review]
much diff very approve

**Pull request merged**